src/haywire/http_request.c

```diff
@@ -143,36 +157,68 @@ int http_request_on_message_begin(http_parser* parser)
+         * the length. When the URL is located later on, it will be guaranteedly contiguous. */
+        url->length += length;
+    } else {
+        url->value = at;
+        url->length = length;
+        /* We've seen the URL start, so we need to pin it to recover it later */
+        http_request_buffer_pin(connection->buffer, url, url->value);
+    }
+
     return 0;
 }
 
+void http_request_save_current_header(http_connection* connection) {
+    hw_string* header_key_copy = hw_strdup(&connection->current_header_key);
+    hw_string* header_value_copy = hw_strdup(&connection->current_header_value);
+
+    /* We duplicate the current header key/value, so we actually need to use the new pointers as pin ids, as
+     * those will be the IDs we'll use later on to locate the headers */
+    http_request_buffer_reassign_pin(connection->buffer, &connection->current_header_key, header_key_copy);
+    http_request_buffer_reassign_pin(connection->buffer, &connection->current_header_value, header_value_copy);
+
+    /* Set headers is going to need to have the values of header key/value, so we need to make sure we get
+     * the pointers pointing at the right place in the buffer */
+    header_key_copy->value = http_request_buffer_locate(connection->buffer, header_key_copy,
+                                                        connection->current_header_key.value);
+
+    header_value_copy->value = http_request_buffer_locate(connection->buffer, header_value_copy,
+                                                          connection->current_header_value.value);
+
+    /* Save last header key/value pair that was read */
+    set_header(connection->request, header_key_copy, header_value_copy);
+}
+
+
 int http_request_on_header_field(http_parser *parser, const char *at, size_t length)
 {
     http_connection* connection = (http_connection*)parser->data;
-    int i = 0;
 
-    if (connection->last_was_value && connection->current_header_key.length > 0)
-    {
-        // Save last read header key/value pair.
-        for (i = 0; i < connection->current_header_key.length; i++)
-        {
-            connection->current_header_key.value[i] = tolower(connection->current_header_key.value[i]);
-        }
+    if (connection->last_was_value && connection->current_header_key.length > 0) {
+        http_request_save_current_header(connection);
 
-        set_header(connection->request, hw_strdup(&connection->current_header_key), hw_strdup(&connection->current_header_value));
-
-        /* Start of a new header */
         connection->current_header_key.length = 0;
+        connection->current_header_value.length = 0;
+    }
+
+    if (connection->current_header_key.length > 0) {
+        /* Another header key chunk, not necessarily contiguous to the other chunks seen before, so we only increment
+         * the length. When the header key is located later on, it will be guaranteedly contiguous. */
```

Comment:
Small nitpick but we talked about rewording our use of contiguous on IRC but also `guaranteedly` is not a word :P Super minor though.